netty/src/test/java/io/grpc/netty/WriteBufferingAndExceptionHandlerTest.java

```diff
@@ -0,0 +1,271 @@
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import io.grpc.Status;
+import io.grpc.Status.Code;
+import io.netty.bootstrap.Bootstrap;
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerAdapter;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.DefaultEventLoop;
+import io.netty.channel.EventLoop;
+import io.netty.channel.local.LocalAddress;
+import io.netty.channel.local.LocalChannel;
+import io.netty.channel.local.LocalServerChannel;
+import java.net.ConnectException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.After;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.DisableOnDebug;
+import org.junit.rules.TestRule;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests for {@link WriteBufferingAndExceptionHandler}.
+ */
+@RunWith(JUnit4.class)
+public class WriteBufferingAndExceptionHandlerTest {
+
+  private static final long TIMEOUT_SECONDS = 10;
+
+  @Rule
+  public final TestRule timeout = new DisableOnDebug(Timeout.seconds(TIMEOUT_SECONDS));
+
+  private final EventLoop group = new DefaultEventLoop();
+  private Channel chan;
+  private Channel server;
+
+  @After
+  public void tearDown() {
+    if (server != null) {
+      server.close();
+    }
+    if (chan != null) {
+      chan.close();
+    }
+    group.shutdownGracefully();
+  }
+
+  @Test
+  public void connectionFailuresPropagated() throws Exception {
+    WriteBufferingAndExceptionHandler handler =
+        new WriteBufferingAndExceptionHandler(new ChannelHandlerAdapter() {});
+    ChannelFuture cf = new Bootstrap()
+        .channel(LocalChannel.class)
+        .handler(handler)
+        .group(group)
+        .register();
+    chan = cf.channel();
+    cf.sync();
+    // Write before connect.  In the event connect fails, the pipeline is torn down and the handler
+    // won't be able to fail the writes with the correct exception.
+    ChannelFuture wf = chan.writeAndFlush(new Object());
+    chan.connect(new LocalAddress("bogus"));
+
+    try {
+      wf.sync();
+      fail();
+    } catch (Exception e) {
+      assertThat(e).isInstanceOf(ConnectException.class);
+      assertThat(e).hasMessageThat().contains("connection refused");
+    }
+  }
+
+  @Test
+  public void channelInactiveFailuresPropagated() throws Exception {
+    WriteBufferingAndExceptionHandler handler =
+        new WriteBufferingAndExceptionHandler(new ChannelHandlerAdapter() {});
+    LocalAddress addr = new LocalAddress("bogus");
```

Comment:
"bogus" worked well in connectionFailuresPropagated because there was no server. For these others, could you rename this, even if it is just "something" or "myfavoritestring"?

Reply:
Done.